ambry-protocol/src/main/java/com/github/ambry/protocol/PutRequest.java

```diff
@@ -175,93 +179,95 @@ private int sizeExcludingBlobAndCrcSize() {
+      nioBuffers = bufferToSend.nioBuffers();
+    }
     long written = 0;
-    try {
-      if (sentBytes < sizeInBytes()) {
-        if (bufferToSend == null) {
-          // this is the first time this method was called, prepare the buffer to send the header and other metadata
-          // (everything except the blob content).
-          prepareBuffer();
-        }
-        // If the header and metadata are not yet written out completely, try and write out as much of it now.
-        if (bufferToSend.hasRemaining()) {
-          written = channel.write(bufferToSend);
-        }
-
-        // If the header and metadata were written out completely (in this call or a previous call),
-        // try and write out as much of the blob now.
-        if (!bufferToSend.hasRemaining() && blob != null) {
-          int totalWrittenBytesForBlob = 0, currentWritten = -1;
-          while (bufferIndex < nioBuffersFromBlob.length && currentWritten != 0) {
-            currentWritten = -1;
-            ByteBuffer byteBuffer = nioBuffersFromBlob[bufferIndex];
-            if (!byteBuffer.hasRemaining()) {
-              // some bytebuffers are zero length, ignore those bytebuffers.
-              bufferIndex ++;
-            } else {
-              currentWritten = channel.write(byteBuffer);
-              totalWrittenBytesForBlob += currentWritten;
-            }
-          }
-          blob.skipBytes(totalWrittenBytesForBlob);
-          written += totalWrittenBytesForBlob;
-          okayToWriteCrc = !blob.isReadable();
-        }
-
-        if (okayToWriteCrc && crcBuf.hasRemaining()) {
-          if (blob != null) {
-            blob.release();
-            blob = null;
-          }
-          written += channel.write(crcBuf);
+    if (sentBytes < sizeInBytes()) {
+      int totalWrittenBytes = 0, currentWritten = -1;
+      while (bufferIndex < nioBuffers.length && currentWritten != 0) {
+        currentWritten = -1;
+        ByteBuffer byteBuffer = nioBuffers[bufferIndex];
+        if (!byteBuffer.hasRemaining()) {
+          // some bytebuffers are zero length, ignore those bytebuffers.
+          bufferIndex++;
+        } else {
+          currentWritten = channel.write(byteBuffer);
+          totalWrittenBytes += currentWritten;
         }
-
-        sentBytes += written;
-      }
-    } catch (Exception e) {
-      if (blob != null) {
-        blob.release();
-        blob = null;
       }
+      bufferToSend.skipBytes(totalWrittenBytes);
+      written += totalWrittenBytes;
     }
+    sentBytes += written;
     return written;
   }
 
+  /**
+   * Override release even if the {@link #content()}'s result is not null. When the {@link #prepareBuffer()} is
+   * not invoked, there will be {@link #blob} and {@link #crcByteBuf} created but not returned by {@link #content()},
+   * and we have to release them.
+   * @return
+   */
   @Override
-  public void release() {
+  public boolean release() {
     if (blob != null) {
       blob.release();
```

Comment:
Http2NetworkClient throws an exception

```
release
[2020-06-26 14:16:03,876] WARN An exception was thrown by com.github.ambry.network.http2.Http2NetworkClient$1.operationComplete() (io.netty.util.concurrent.DefaultPromise)
io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.util.internal.ReferenceCountUpdater.toLiveRealRefCnt(ReferenceCountUpdater.java:74)
	at io.netty.util.internal.ReferenceCountUpdater.release(ReferenceCountUpdater.java:138)
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:100)
	at com.github.ambry.protocol.PutRequest.release(PutRequest.java:269)
```

Netty write decrease the ref cnt, but it doesn't set the variable to null. So we hit above exception.

Reply:
I purposely didn't set the content to null so we don't release it multiple times. Where does it get released before? Can we figure that out?

Reply:
In Http2NetworkClient::sendAndPoll()
```
                  // release related bytebuf
                  requestInfo.getRequest().release();
```
This is called in the success case, but with your PR, this no longer needed.